core/platform/auth/firebase_auth_services.py

```diff
@@ -0,0 +1,203 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Service layer for handling user-authentication via Firebase.
+
+Oppia depends on OpenID Connect 1.0 to handle user authentication. We use
+[Firebase authentication](https://firebase.google.com/docs/auth) to do the
+heavy-lifting, especially for securely storing user credentials and associations
+to their identity providers. This helps us minimize our contact with these
+private details.
+
+Terminology:
+    OpenID Connect 1.0 (OIDC):
+        A simple identity layer on top of the OAuth 2.0 protocol. It is a
+        specification (i.e., a strict set of algorithms, data structures, and
+        rules) that defines how two parties must share data about a user
+        securely, on that user's behalf.
+    OAuth 2.0 (OAuth):
+        The industry-standard protocol for authorization. It enables a
+        third-party application to obtain limited access to an HTTP service on
+        behalf of a resource owner.
+    Claim:
+        A piece of information about a user (name, address, phone number, etc.)
+        that has been encrypted and digitally signed.
+    JSON Web Token (JWT):
+        A compact and URL-safe protocol primarily designed to send Claims
+        between two parties. Claims are organized into JSON objects mapping
+        "Claim Names" to "Claim Values".
+    Identity provider:
+        An entity that creates, maintains, and manages identity information and
+        provides authentication services. Such services rely on JWTs to send
+        identity information. Examples of identity providers include: Google,
+        Facebook, Email verification links, and Text message SMS codes.
+    Subject Identifier:
+        A Claim that can uniquely identify a user. It is locally unique and
+        never reassigned with respect to the provider who issued it.
+        For example: `24400320` or `AItOawmwtWwcT0k51BayewNvutrJUqsvl6qs7A4`.
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import logging
+
+from core.platform import models
+import python_utils
+
+import firebase_admin
+from firebase_admin import auth as firebase_auth
+
+(auth_models,) = models.Registry.import_models([models.NAMES.auth])
+
+
+def _ensure_firebase_is_initialized():
+    """Initializes the Firebase Admin SDK."""
+    try:
+        firebase_admin.get_app()
+    except Exception:
```

Comment:
Should we tighten this to more specific `Exception` that is directly raise by `get_app`, we are trying to limit the usage of general `Exception`.

Reply:
Sure, done.